tensornetwork/sparse/lanczos.py

```diff
@@ -0,0 +1,315 @@
+    Args:
+      dot_product: A function which implements a dot-product between two vectors
+        Allowed signatures: `dot_product(x, y)`, `dot_product(x, y, backend)`
+          and `dot_product(x, y, backend=default)` with `default` 'numpy', 
+          'tensorflow', 'pytorch', 'jax' or 'shell'.
+         `result = `dot_product` should return `result` of the same type as `x` 
+          (i.e. a `Tensor` object)
+          Note that any default value for argument `backend` of `dot_product` 
+          is overridden by the value of the `backend` argument of 
+          `ScalarProduct.__init__`
+      dtype: The data-type of the `ScalarProduct`.
+      backend: The backend to be used with this `ScalarProduct`. 
+        `backend` can be 'numpy', 'tensorflow', 'pytorch', 'jax' or 'shell'
+    Returns: 
+      None
+    Raises:
+      ValueError if `dot_product` signature is other than the above mentioned.
+    """
+
+    self.backend = backend_factory.get_backend(backend, dtype)
+    args = inspect.getargspec(dot_product)
+    if len(args[0]) > 3:
+      raise ValueError(
+          '`dot_product` can have at most 3 arguments; found args = {}'.format(
+              args[0]))
+    if (args[0][0] == 'backend') or (args[0][1] == 'backend'):
+      raise ValueError(
+          "Found `backend` as first or second argument of `dot_product`. "
+          "It can only be the third argument!")
+    if args[3] and (len(args[3]) != 1):
+      N = len(args[0])
+      params = [args[0][N - 1 - n] for n in range(len(args[3]))]
+      defaults = [args[3][n] for n in reversed(range(len(args[3])))]
+      raise ValueError(
+          "The only allowed argument to `dot_product` with defaults is "
+          "`backend`. Found arguments {} with default values {}!".format(
+              params[::-1], defaults[::-1]))
+
+    if args[3] and (len(args[3]) == 1):
+      if args[3][0] not in ('tensorflow', 'numpy', 'jax', 'pytorch', 'shell'):
+        raise ValueError(
+            "wrong default '{}' for argument `{}` of `dot_product`. "
+            "Only allowed values are 'numpy', 'tensorflow', 'pytorch', "
+            "'jax' and 'shell'".format(args[3][0], args[0][1]))
+      if args[3][0] != self.backend.name:
+        warnings.warn(
+            "default value of parameter `{0}` = '{1}' of `dot_product` is "
+            "different from ScalarProduct.backend.name='{2}'."
+            " Overriding the default to `{0}` = '{2}'".format(
+                args[0][1], args[3][0], self.backend.name))
+
+    if len(args[0]) == 2:  #dot_product takes only one argument
+      self.dot_product = dot_product
+    else:  #dot_product takes two arguments (x, backend)
+
+      def _dot_product(x, y):
+        return dot_product(x, y, self.backend.name)
+
+      self.dot_product = _dot_product
+
+  @property
+  def dtype(self):
+    return self.backend.dtype
+
+  def __call__(self, x, y):
+    return self.dot_product(x, y)
+
+
+def eigsh_lanczos(A: LinearOperator,
```

Comment:
I don't understand why you needed to build this. Why would this implementation of an eigen value solver be better than the built in ones?

Reply:
Until this function works under a `jax.jit`, I can't pull this in.